src/shims/foreign_items.rs

```diff
@@ -102,6 +118,78 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
     }
 
+    /// Lookup the body of a function that has `link_name` as the symbol name.
+    fn lookup_exported_symbol(
+        &mut self,
+        abi: Abi,
+        link_name: Symbol,
+    ) -> InterpResult<'tcx, Option<&'mir mir::Body<'tcx>>> {
+        let this = self.eval_context_mut();
+        let tcx = this.tcx.tcx;
+        let load_mir = |this: &crate::MiriEvalContext<'_, 'tcx>, instance: &ty::Instance<'tcx>| {
+            check_abi(abi, instance.ty(tcx, ty::ParamEnv::reveal_all()).fn_sig(tcx).abi())?;
+            this.load_mir(instance.def, None)
+        };
+
+        // If the result was cached, just perform ABI check and load MIR.
+        if let Some(instance) = this.machine.exported_symbols_cache.get(&link_name) {
+            return load_mir(this, instance).map(Some);
+        }
+
+        // Find it if it was not cached.
+        let mut instance_and_crate: Option<(ty::Instance<'_>, CrateNum)> = None;
+        for cnum in iter::once(LOCAL_CRATE)
+            .chain(
+                tcx.dependency_formats(LOCAL_CRATE)
+                    .iter()
+                    .find(|(crate_type, _)| *crate_type == CrateType::Executable)
+                    .expect("interpreting a non-executable crate")
+                    .1
```

Comment:
I don't know this `dependency_formats` query... does that really list all *transitive* dependencies?

Either way, please `let`-bind the first part of this expression, until `.expect` -- that seems to be a meaningful part on its own.

Reply:
> does that really list all transitive dependencies?

I added a test for this, and looks like it does.

Reply:
I'd like to understand this `dependency_formats` query we rely on before considering this problem resolved. The only docs I found so far is

> get the linkage format of all dependencies

so probably we'll have to look at the implementation...

Reply:
It's implemented in https://github.com/rust-lang/rust/blob/94ecdfd115a353f65d4b21056b23c7074339be6e/compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs#L373-L376, and this is what it does to collect all crates: https://github.com/rust-lang/rust/blob/94ecdfd115a353f65d4b21056b23c7074339be6e/compiler/rustc_metadata/src/dependency_format.rs#L173-L174

Reply:
Okay that does sound pretty transitive. I presume this is what the compiler uses as the underlying logic for linking.

Please add a comment explaining that `dependency_formats` seems to include all the transitive information needed to link a crate, and thus is exactly what we need here. Then I'm happy. :)